platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/DefaultBuildController.java

```diff
@@ -108,69 +98,43 @@ public BuildResult<?> getModel(@Nullable Object target, ModelIdentifier modelIde
         if (cancellationToken.isCancellationRequested()) {
             throw new BuildCancelledException(String.format("Could not build '%s' model. Build cancelled.", modelIdentifier.getName()));
         }
-        ToolingModelScope scope = getTarget(target, modelIdentifier, parameter != null);
 
-        Object model;
+        BuildTreeModelTarget scopedTarget = getTarget(target);
         try {
-            if (parameter == null) {
-                model = scope.getModel(modelIdentifier.getName(), null);
-            } else {
-                model = scope.getModel(modelIdentifier.getName(), parameterCarrierFactory.createCarrier(parameter));
-            }
+            Object model = controller.getModel(scopedTarget, modelIdentifier.getName(), parameter);
+            return new ProviderBuildResult<>(model);
         } catch (UnknownModelException e) {
             throw (InternalUnsupportedModelException) new InternalUnsupportedModelException().initCause(e);
         }
-
-        return new ProviderBuildResult<>(model);
-    }
-
-    @Override
-    public boolean getCanQueryProjectModelInParallel(Class<?> modelType) {
-        return controller.queryModelActionsRunInParallel();
-    }
-
-    @Override
-    public <T> List<T> run(List<Supplier<T>> actions) {
-        assertCanQuery();
-        return controller.runQueryModelActions(actions);
     }
 
-    private ToolingModelScope getTarget(@Nullable Object target, ModelIdentifier modelIdentifier, boolean parameter) {
+    @Nullable
+    private static BuildTreeModelTarget getTarget(@Nullable Object target) {
         if (target == null) {
-            return controller.locateBuilderForDefaultTarget(modelIdentifier.getName(), parameter);
+            return null;
```

Comment:
🔍 Methods that accept `null` to return `null` have a particular smell to me. I'd make the parameter non-nullable and move the decision to the call-site. That way this method's return value also become non-nullable.

Reply:
I agree about the null => null being a code smell, I'm not sure about making the argument non-nullable.

The method argument here is whatever the TAPI client received from the user, and the purpose of this method is to map this target into another value space of `BuildTreeModelTarget?`. In the current state of the PR, the default target is modeled as `null`.

Applying the "sealed" hierarchy approach then makes more sense, having the `Default` target represented explicitly in the `BuildTreeModelTarget`.